prow/tide/history/history.go

```diff
@@ -33,14 +38,84 @@ import (
 // Mock out time for unit testing.
 var now = time.Now
 
+type storageObject interface {
+	NewReader(context.Context) (io.ReadCloser, error)
+	NewWriter(context.Context) io.WriteCloser
+}
+
+type gcsStorageObject struct {
+	*storage.ObjectHandle
+}
+
+func (o gcsStorageObject) NewReader(c context.Context) (io.ReadCloser, error) {
+	return o.ObjectHandle.NewReader(c)
+}
+func (o gcsStorageObject) NewWriter(c context.Context) io.WriteCloser {
+	return o.ObjectHandle.NewWriter(c)
+}
+
 // History uses a `*recordLog` per pool to store a record of recent actions that
 // Tide has taken. Using a log per pool ensure that history is retained
 // for inactive pools even if other pools are very active.
 type History struct {
 	logs map[string]*recordLog
 	sync.Mutex
-
 	logSizeLimit int
+
+	gcsObj *storage.ObjectHandle
+}
+
+func readHistory(maxRecordsPerKey int, obj storageObject) (map[string]*recordLog, error) {
+	reader, err := obj.NewReader(context.Background())
+	if err == storage.ErrObjectNotExist {
+		// No history exists yet. This is not an error.
+		return map[string]*recordLog{}, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error opening GCS object: %v", err)
+	}
+	defer func() {
+		if err := reader.Close(); err != nil {
+			logrus.WithError(err).Error("Error closing GCS object reader.")
+		}
+	}()
+	raw, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, fmt.Errorf("error reading GCS object: %v", err)
+	}
+	var recordsMap map[string][]*Record
+	if err := json.Unmarshal(raw, &recordsMap); err != nil {
+		return nil, fmt.Errorf("error unmarshaling GCS object: %v", err)
+	}
+
+	// Load records into a new recordLog map.
+	recLogMap := make(map[string]*recordLog, len(recordsMap))
+	for poolKey, records := range recordsMap {
```

Comment:
This might be a nice thing to break out and test, without also involving JSON parsing etc

Reply:
Its only 10 lines and already well tested by the test cases. I don't see much value in refactoring it to test without the JSON parsing.
Are you ok with me leaving this as is?